File: languages/rust/uhppoted/src/udp.rs
use std::net::UdpSocket;
use std::sync::RwLock;
use std::time::Duration;

use async_std::future;
use futures;
use lazy_static::lazy_static;

use super::error;

const READ_TIMEOUT: Duration = Duration::from_millis(5000);
const WRITE_TIMEOUT: Duration = Duration::from_millis(1000);
const TIMEOUT: Duration = Duration::from_millis(2500);

lazy_static! {
    static ref BIND_ADDR: RwLock<String> = RwLock::new("0.0.0.0:0".to_string());
    static ref BROADCAST_ADDR: RwLock<String> = RwLock::new("255.255.255.255:60000".to_string());
    static ref DEBUG: RwLock<bool> = RwLock::new(false);
}

type ReadFn = fn(socket: UdpSocket) -> Result<Vec<[u8; 64]>, error::Error>;

pub fn set_bind_addr(addr: &str) {
    if let Ok(mut guard) = BIND_ADDR.write() {
        *guard = addr.to_string();
    }
}

pub fn set_broadcast_addr(addr: &str) {
    if let Ok(mut guard) = BROADCAST_ADDR.write() {
        *guard = addr.to_string();
    }
}

pub fn set_debug(enabled: bool) {
    if let Ok(mut guard) = DEBUG.write() {
        *guard = enabled;
    }
}

pub fn send(packet: &[u8; 64], f: ReadFn) -> Result<Vec<[u8; 64]>, error::Error> {
    let bind = BIND_ADDR.read()?;
    let broadcast = BROADCAST_ADDR.read()?;
    let socket = UdpSocket::bind(bind.as_str())?;

    dump(&packet);

    socket.set_write_timeout(Some(WRITE_TIMEOUT))?;
    socket.set_read_timeout(Some(READ_TIMEOUT))?;
    socket.set_broadcast(true)?;
    socket.send_to(packet, broadcast.as_str())?;

    return f(socket);
}

pub fn read(socket: UdpSocket) -> Result<Vec<[u8; 64]>, error::Error> {
    let f = future::timeout(TIMEOUT, async {
        let sock = async_std::net::UdpSocket::from(socket);
        let mut buffer = [0u8; 1024];

        let result: Result<[u8; 64], error::Error> = loop {
            match sock.recv(&mut buffer).await {
                Ok(64) => {
                    dump(buffer[..64].try_into()?);
                    break Ok(buffer[..64].try_into()?);
                }

                Err(e) => break Err(error::Error::from(e)),
                _ => continue,
            }
        };

        return result;
    });

    let g = async {
        match f.await {
            Ok(v) => match v {
                Ok(reply) => return Ok(vec![reply]),
                Err(e) => return Err(e),
            },

            Err(e) => return Err(error::Error::from(e)),
        }
    };

    match futures::executor::block_on(g) {
        Ok(replies) => return Ok(replies),
        Err(e) => return Err(e),
    }
}

pub fn read_all(socket: UdpSocket) -> Result<Vec<[u8; 64]>, error::Error> {
    let replies = RwLock::<Vec<[u8; 64]>>::new(vec![]);

    let f = future::timeout(Duration::from_millis(2500), async {
        let sock = async_std::net::UdpSocket::from(socket);
        let mut buffer = [0u8; 1024];

        let result: Result<(), error::Error> = loop {
            match sock.recv(&mut buffer).await {
                Ok(64) => {
                    dump(buffer[..64].try_into()?);
                    replies.write().unwrap().push(buffer[..64].try_into()?);
                }

                Err(e) => break Err(error::Error::from(e)),
                _ => continue,
            }
        };

        return result;
    });

    let g = async {
        match f.await {
            Ok(v) => match v {
                Ok(_) => return Ok(replies.read().unwrap().to_vec()),
                Err(e) => return Err(error::Error::from(e)),
            },

            Err(_) => return Ok(replies.read().unwrap().to_vec()),
        }
    };

    match futures::executor::block_on(g) {
        Ok(replies) => return Ok(replies),
        Err(e) => return Err(e),
    }
}

pub fn read_none(_: UdpSocket) -> Result<Vec<[u8; 64]>, error::Error> {
    return Ok(vec![]);
}

fn dump(packet: &[u8; 64]) {
    let debug = if let Ok(guard) = DEBUG.read() {
        *guard
    } else {
        false
    };

    if debug {
        for i in 0..4 {
            let offset = i * 16;
            let u = &packet[offset..offset + 8];
            let v = &packet[offset + 8..offset + 16];

            let p = format!(
                "{:02x} {:02x} {:02x} {:02x} {:02x} {:02x} {:02x} {:02x}",
                u[0], u[1], u[2], u[3], u[4], u[5], u[6], u[7]
            );
            let q = format!(
                "{:02x} {:02x} {:02x} {:02x} {:02x} {:02x} {:02x} {:02x}",
                v[0], v[1], v[2], v[3], v[4], v[5], v[6], v[7]
            );

            println!("   {offset:08x}  {p}  {q}");
        }

        println!();
    }
}

File: languages/go/uhppote/udp.go
package uhppote

import (
    "fmt"
    "net"
    "net/netip"
    "time"
)

const READ_TIMEOUT = 5000 * time.Millisecond
const WRITE_TIMEOUT = 1000 * time.Millisecond

var bindAddr *net.UDPAddr = resolve("0.0.0.0:0")
var destAddr *net.UDPAddr = resolve("255.255.255.255:60000")
var debug bool = false

func SetBindAddr(addr string) {
    address := netip.MustParseAddr(addr)
    addrPort := netip.AddrPortFrom(address, 0)
    bindAddr = net.UDPAddrFromAddrPort(addrPort)
}

func SetDestAddr(addr string) {
    destAddr = resolve(addr)
}

func SetDebug(enabled bool) {
    debug = enabled
}

func send(request []byte, f func(*net.UDPConn) ([][]byte, error)) ([][]byte, error) {
    dump(request)

    socket, err := net.ListenUDP("udp", bindAddr)
    if err != nil {
        return nil, err
    }

    defer socket.Close()

    if err := socket.SetWriteDeadline(time.Now().Add(WRITE_TIMEOUT)); err != nil {
        return nil, err
    }

    if err := socket.SetReadDeadline(time.Now().Add(READ_TIMEOUT)); err != nil {
        return nil, fmt.Errorf("Failed to set UDP read timeout [%v]", err)
    }

    if _, err := socket.WriteToUDP(request, destAddr); err != nil {
        return nil, err
    }

    return f(socket)
}

func readAll(socket *net.UDPConn) ([][]byte, error) {
    reply := make(chan []byte)
    e := make(chan error)

    go func() {
        buffer := make([]byte, 1024)

        for {
            if N, _, err := socket.ReadFromUDP(buffer); err != nil {
                e <- err
            } else if N == 64 {
                m := make([]byte, 64)
                copy(m, buffer[0:N])
                reply <- m
            }
        }
    }()

    replies := [][]byte{}
    wait := time.After(2500 * time.Millisecond)

    for {
        select {
        case m := <-reply:
            dump(m)
            replies = append(replies, m)

        case <-wait:
            return replies, nil

        case err := <-e:
            return nil, err
        }
    }
}

func readNone(socket *net.UDPConn) ([][]byte, error) {
    return [][]byte{}, nil
}

func read(socket *net.UDPConn) ([][]byte, error) {
    reply := make(chan []byte)
    e := make(chan error)

    go func() {
        buffer := make([]byte, 1024)

        for {
            if N, _, err := socket.ReadFromUDP(buffer); err != nil {
                e <- err
            } else if N == 64 {
                reply <- buffer[0:N]
            }
        }
    }()

    timeout := time.After(5000 * time.Millisecond)

    for {
        select {
        case m := <-reply:
            dump(m)
            return [][]byte{m}, nil

        case <-timeout:
            return nil, fmt.Errorf("timeout")

        case err := <-e:
            return nil, err
        }
    }
}

func resolve(address string) *net.UDPAddr {
    addr := netip.MustParseAddrPort(address)

    return net.UDPAddrFromAddrPort(addr)
}

func dump(packet []byte) {
    if debug {
        hex := "%02x %02x %02x %02x %02x %02x %02x %02x"

        for i := 0; i < 4; i++ {
            offset := i * 16
            u := packet[offset : offset+8]
            v := packet[offset+8 : offset+16]

            p := fmt.Sprintf(hex, u[0], u[1], u[2], u[3], u[4], u[5], u[6], u[7])
            q := fmt.Sprintf(hex, v[0], v[1], v[2], v[3], v[4], v[5], v[6], v[7])

            fmt.Printf("   %08x  %v  %v\n", offset, p, q)
        }

        fmt.Println()
    }
}
